File: ai.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <unistd.h>

/*constant-------------------------------------------------------*/
#define COLUMN 4
#define ROW 4
#define DEBUG true


/*initial funtion definition-------------------------------------*/
void sendCommand();

void goUp();
void goDown();
void goLeft();
void goRight();

/*public variables-----------------------------------------------*/
int received_data[ROW * COLUMN + 1];

/*main program---------------------------------------------------*/

int main()
{
    while (true) {
        //scanfでは空白で区切られる
        for (int i = 0; i < ROW * COLUMN + 1; i++) {
            scanf("%d", &received_data[i]);
        }
        printf("down\n");
        usleep(1000);

        // 盤面状態の数字が返ってきてないときエラーを吐いて次のループへ
        if (received_data[ROW * COLUMN] != 0 && received_data[ROW * COLUMN] != 1 && received_data[ROW * COLUMN] != 2) {
            printf("An error has occured in 2048/2048.rb\n");
            continue;
        }
        //status 2を取得した時ゲームオーバー
        if (received_data[ROW * COLUMN] == 2) {
            printf("Game Over.\n");
            exit(1);
            continue;
        }

        /*AI program-----------------------------------------------------*/
        for (int i = 0; i < COLUMN; i++) {
            for (int j = 0; j < ROW; j++) {
                printf("%d", received_data[i * 4 + j]);
            }
            printf("\n");
        }
        printf("gamestatus %d\n", received_data[ROW * COLUMN]);

        if (received_data[ROW * COLUMN] == 1) {
            goDown();
        } else {
            static int debug_i = 0;
            switch (debug_i) {
            case 0:
                goUp();
                break;
            case 1:
                goRight();
                break;
            case 2:
                goUp();
                break;
            case 3:
                goLeft();
                break;
            default:
                printf("Invalid Command\n");
                break;
            }
            if (debug_i == 3) {
                debug_i = 0;
            } else {
                debug_i++;
            }
        }
    }
    return 0;
}

void goUp()
{
    printf("up\n");
}
void goDown()
{
    printf("down\n");
}
void goLeft()
{
    printf("left\n");
}
void goRight()
{
    printf("right\n");
}

File: ruby/ai.rb
#!/usr/bin/ruby
STDOUT.sync = true
require 'pp'
# Constant--------------------------------------------------------
COLUMN = 4
ROW = 4
DEBUG = true

# Public Variables------------------------------------------------
@debug_i = 0;

# Preparing the data table----------------------------------------
@board_data = Array.new(ROW){Array.new(COLUMN, 0)}

# Funtion definition----------------------------------------------

def getTileInfo(received_data)
    # Splitting the line by Space.
    # Ignoring multi-space.
    numbers = received_data.split(/\s+/)
    for i in 0..(ROW - 1)
        for j in 0..(COLUMN - 1)
            @board_data[i][j] = numbers[i * COLUMN + j]
        end
    end
end

def getGameStatus(received_data)
    numbers = received_data.split(/\s+/)
    numbers[ROW * COLUMN]
end

def goUp()
    puts "up"
end
def goDown()
    puts "down"
end
def goLeft()
    puts "left"
end
def goRight()
    puts "right"
end


#public variables-------------------------------------------------


#main program-----------------------------------------------------

begin
    while true
        received_data = gets.chomp

        if received_data == ""
            puts "an error has occured in 2048/2048.rb"
            next
        end
        # gameover judge
        if getGameStatus(received_data) == 2
            puts "Game Over"
            exit 1
            next
        end

        #AI program-------------------------------------------------------
        getTileInfo(received_data)
        for i in 0..(COLUMN - 1)
            for j in 0..(ROW - 1)
                print @board_data[i][j] + " "
            end
            puts ""
        end
        puts "GameStatus: #{getGameStatus(received_data)}"

        case @debug_i
        when 0 then
            goUp()
        when 1 then
            goRight()
        when 2 then
            goDown()
        when 3 then
            goLeft()
        else
            puts "Invalid Command"
            puts __LINE__
        end
        if @debug_i == 3
            @debug_i = 0
        else
            @debug_i += 1
        end
    end
rescue => general_exception
    puts "General Error: #{general_exception.message}"
    pp general_exception
end
